logstash-core/src/main/java/org/logstash/execution/QueueReadClientBase.java

```diff
@@ -0,0 +1,180 @@
+        eventMetricOut = LongCounter.fromRubyBase(metric, OUT_KEY);
+        eventMetricFiltered = LongCounter.fromRubyBase(metric, FILTERED_KEY);
+        eventMetricTime = LongCounter.fromRubyBase(metric, DURATION_IN_MILLIS_KEY);
+        return this;
+    }
+
+    @JRubyMethod(name = "set_pipeline_metric", required = 1)
+    public IRubyObject setPipelineMetric(final ThreadContext context, IRubyObject metric) {
+        pipelineMetricOut = LongCounter.fromRubyBase(metric, OUT_KEY);
+        pipelineMetricFiltered = LongCounter.fromRubyBase(metric, FILTERED_KEY);
+        pipelineMetricTime = LongCounter.fromRubyBase(metric, DURATION_IN_MILLIS_KEY);
+        return this;
+    }
+
+    @JRubyMethod(name = "set_batch_dimensions")
+    public IRubyObject rubySetBatchDimensions(final ThreadContext context, IRubyObject batchSize,
+                                              IRubyObject waitForMillis) {
+        setBatchDimensions(((RubyNumeric) batchSize).getIntValue(),
+                ((RubyNumeric) waitForMillis).getIntValue());
+        return this;
+    }
+
+    public void setBatchDimensions(int batchSize, int waitForMillis) {
+        this.batchSize = batchSize;
+        this.waitForNanos = TimeUnit.NANOSECONDS.convert(waitForMillis, TimeUnit.MILLISECONDS);
+        this.waitForMillis = waitForMillis;
+    }
+
+    @JRubyMethod(name = "empty?")
+    public IRubyObject rubyIsEmpty(final ThreadContext context) {
+        return context.runtime.newBoolean(isEmpty());
+    }
+
+    @JRubyMethod(name = "close")
+    public void rubyClose(final ThreadContext context) {
+        try {
+            close();
+        } catch (IOException e) {
+            throw RubyUtil.newRubyIOError(context.runtime, e);
+        }
+    }
+
+    @JRubyMethod(name = "read_batch")
+    public IRubyObject rubyReadBatch(final ThreadContext context) throws InterruptedException {
+        return JavaObject.wrap(context.runtime, readBatch());
+    }
+
+    @Override
+    public void closeBatch(QueueBatch batch) throws IOException {
+        batch.close();
+        inflightBatches.remove(Thread.currentThread().getId());
+        Long startTime = inflightClocks.remove(Thread.currentThread().getId());
+        if (startTime != null && batch.filteredSize() > 0) {
+            // stop timer and record metrics iff the batch is non-empty.
+            long elapsedTimeMillis = (System.nanoTime() - startTime) / 1_000_000;
+            eventMetricTime.increment(elapsedTimeMillis);
+            pipelineMetricTime.increment(elapsedTimeMillis);
+        }
+    }
+
+    // currently used only in the original pipeline and rspec tests
+    @JRubyMethod(name = "close_batch")
+    public void rubyCloseBatch(final ThreadContext context, IRubyObject batch) throws IOException {
+
+        // Only the Ruby pipeline uses JavaProxy instances. Once that is fully deprecated, this
+        // type checking can be removed.
+        QueueBatch queueBatch = null;
```

Comment:
@danhermann nit: no need for the `null` here imo. Also, you could make this just `final QueueBatch queueBatch` + extract this logic to a `private static` method since we do this in at least 2 places right? :)